struct forward declaration

The following program `sample.go` triggers a panic:

```go
package main
  
type T struct {
    f func(t *T1)
    y *xxx
}

type T1 struct {
    T
}

func f(o *T) { println("in f") }

var x = &T1{}

type xxx struct{}

func main() {
    println("ok")
}
```

Expected result:
```console
$ go run ./sample.go
ok
```

Got:
```console
$ yaegi ./sample.go
sample.go:5:8: incomplete type T
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc000390000, 0x40)
        /home/marc/sdk/go1.14.1/src/runtime/debug/stack.go:24 +0x9d
github.com/containous/yaegi/interp.(*Interpreter).Eval.func1(0xc0000b9e10)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:320 +0xbc
panic(0xd6f3c0, 0xc00028a920)
        /home/marc/sdk/go1.14.1/src/runtime/panic.go:967 +0x166
github.com/containous/yaegi/interp.(*itype).refType(0xc000377790, 0xc0000b9508, 0xc000286000, 0x2, 0xc0000b95c0)
        /home/marc/go/src/github.com/containous/yaegi/interp/type.go:923 +0xea5
github.com/containous/yaegi/interp.(*itype).refType(0xc000377ad0, 0xc0000b9508, 0xc000286000, 0x2, 0x158b500)
        /home/marc/go/src/github.com/containous/yaegi/interp/type.go:968 +0x7c2
github.com/containous/yaegi/interp.(*itype).refType(0xc000377e10, 0xc0000b9508, 0xc000282100, 0x19, 0x20)
        /home/marc/go/src/github.com/containous/yaegi/interp/type.go:961 +0xaf3
github.com/containous/yaegi/interp.(*itype).TypeOf(0xc000377e10, 0xc000377e10, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/type.go:986 +0x96
github.com/containous/yaegi/interp.(*itype).frameType(0xc000377e10, 0xc00007e0d0, 0x155c640)
        /home/marc/go/src/github.com/containous/yaegi/interp/type.go:1008 +0xec
github.com/containous/yaegi/interp.(*scope).add(0xc00029c3c0, 0xc000377e10, 0x3)
        /home/marc/go/src/github.com/containous/yaegi/interp/scope.go:156 +0x49
github.com/containous/yaegi/interp.(*Interpreter).gta.func1(0xc00038aa00, 0xc00037a000)
        /home/marc/go/src/github.com/containous/yaegi/interp/gta.go:83 +0x48a
github.com/containous/yaegi/interp.(*node).Walk(0xc00038aa00, 0xc0000b9be8, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:184 +0xc9
github.com/containous/yaegi/interp.(*node).Walk(0xc00038a900, 0xc0000b9be8, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:188 +0x70
github.com/containous/yaegi/interp.(*node).Walk(0xc000294100, 0xc0000b9be8, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:188 +0x70
github.com/containous/yaegi/interp.(*Interpreter).gta(0xc00029a180, 0xc000294100, 0xc000286080, 0x4, 0x7ffcaa6316c3, 0x15, 0x0, 0xc000288280, 0x0, 0x0, ...)
        /home/marc/go/src/github.com/containous/yaegi/interp/gta.go:18 +0x153
github.com/containous/yaegi/interp.(*Interpreter).gtaRetry(0xc00029a180, 0xc0000b9db8, 0x1, 0x1, 0xc000286080, 0x4, 0x7ffcaa6316c3, 0x15, 0x0, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/gta.go:232 +0x14a
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc00029a180, 0xc000380000, 0xca, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:338 +0x194
main.main()
        /home/marc/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:94 +0x548
```

--- a/_test/struct39.go
+++ b/_test/struct39.go
@@ -1,0 +1,22 @@
+package main
+
+type T struct {
+	t *T1
+	y *xxx
+}
+
+type T1 struct {
+	T
+}
+
+var x = &T1{}
+var t = &T{}
+
+type xxx struct{}
+
+func main() {
+	println("ok")
+}
+
+// Output:
+// ok

--- a/_test/struct40.go
+++ b/_test/struct40.go
@@ -1,0 +1,23 @@
+package main
+
+type T struct {
+	t *T1
+	y *xxx
+}
+
+type T1 struct {
+	T
+}
+
+func f(t *T) { println("in f") }
+
+var x = &T1{}
+
+type xxx struct{}
+
+func main() {
+	println("ok")
+}
+
+// Output:
+// ok

--- a/interp/gta.go
+++ b/interp/gta.go
@@ -63,7 +63,7 @@ func (interp *Interpreter) gta(root *node, rpath, pkgID string) ([]*node, error)
 					}
 					val = src.rval
 				}
-				if typ.incomplete {
+				if !typ.isComplete() {
 					// Come back when type is known.
 					revisit = append(revisit, n)
 					return false
@@ -94,7 +94,7 @@ func (interp *Interpreter) gta(root *node, rpath, pkgID string) ([]*node, error)
 				if n.typ, err = nodeType(interp, sc, n.child[l]); err != nil {
 					return false
 				}
-				if n.typ.incomplete {
+				if !n.typ.isComplete() {
 					// Come back when type is known.
 					revisit = append(revisit, n)
 					return false
@@ -139,7 +139,7 @@ func (interp *Interpreter) gta(root *node, rpath, pkgID string) ([]*node, error)
 				// Add a function symbol in the package name space
 				sc.sym[n.child[1].ident] = &symbol{kind: funcSym, typ: n.typ, node: n, index: -1}
 			}
-			if n.typ.incomplete {
+			if !n.typ.isComplete() {
 				revisit = append(revisit, n)
 			}
 			return false
@@ -206,7 +206,7 @@ func (interp *Interpreter) gta(root *node, rpath, pkgID string) ([]*node, error)
 				n.typ.method = append(n.typ.method, sc.sym[typeName].typ.method...)
 			}
 			sc.sym[typeName].typ = n.typ
-			if n.typ.incomplete {
+			if !n.typ.isComplete() {
 				revisit = append(revisit, n)
 			}
 			return false

--- a/interp/type.go
+++ b/interp/type.go
@@ -654,6 +654,43 @@ func (t *itype) finalize() (*itype, error) {
 	return t, err
 }
 
+// isComplete returns true if type definition is complete.
+func (t *itype) isComplete() bool { return isComplete(t, map[string]bool{}) }
+
+func isComplete(t *itype, visited map[string]bool) bool {
+	if t.incomplete {
+		return false
+	}
+	if visited[t.name] {
+		return !t.incomplete
+	}
+	if t.name != "" {
+		visited[t.name] = true
+	}
+	switch t.cat {
+	case aliasT, arrayT, chanT, ptrT:
+		return isComplete(t.val, visited)
+	case funcT:
+		complete := true
+		for _, a := range t.arg {
+			complete = complete && isComplete(a, visited)
+		}
+		for _, a := range t.ret {
+			complete = complete && isComplete(a, visited)
+		}
+		return complete
+	case interfaceT, structT:
+		complete := true
+		for _, f := range t.field {
+			complete = complete && isComplete(f.typ, visited)
+		}
+		return complete
+	case nilT:
+		return false
+	}
+	return true
+}
+
 // Equals returns true if the given type is identical to the receiver one.
 func (t *itype) equals(o *itype) bool {
 	switch ti, oi := isInterface(t), isInterface(o); {

